File: DP2122/DP2122_reader.js
// Function to process raw hex data
function processSensorData(rawData) {
  const SCALING_FACTOR = 0.1;

  // Extract speed (first 4 bytes)
  const speedRaw = parseInt(rawData.slice(0, 8), 16);
  const speed = speedRaw * SCALING_FACTOR;

  // Extract device status (next byte)
  const deviceStatus = parseInt(rawData.slice(8, 10), 16);

  // Extract OUT1 and OUT2 statuses (last byte)
  const digitalStatus = parseInt(rawData.slice(10, 12), 16);
  const out1Status = Boolean(digitalStatus & 0x01);
  const out2Status = Boolean((digitalStatus >> 1) & 0x01);

  return {
    speed: `${speed} Hz`,
    deviceStatus,
    out1Status,
    out2Status,
  };
}

// Function to handle HTTP request
async function handleSensorRequest(req, res) {
  if (req.method !== "POST") {
    res.statusCode = 405; // Method Not Allowed
    res.end(JSON.stringify({ error: "Method not allowed. Use POST." }));
    return;
  }

  try {
    let body = "";
    req.on("data", (chunk) => {
      body += chunk;
    });

    req.on("end", () => {
      const { rawData } = JSON.parse(body); // Assume JSON payload { "rawData": "..." }

      if (!rawData || typeof rawData !== "string" || rawData.length < 12) {
        throw new Error("Invalid rawData format.");
      }

      const processedData = processSensorData(rawData);

      res.statusCode = 200;
      res.setHeader("Content-Type", "application/json");
      res.end(
        JSON.stringify({
          timestamp: new Date().toISOString(),
          rawData,
          ...processedData,
        })
      );
    });
  } catch (error) {
    res.statusCode = 400; // Bad Request
    res.end(JSON.stringify({ error: error.message }));
  }
}

// HTTP Server setup
const http = require("http");

const server = http.createServer((req, res) => {
  handleSensorRequest(req, res);
});

const PORT = 3000;
server.listen(PORT, () => {
  console.log(`Server is listening on port ${PORT}`);
});

File: DP2122/DP2122_reader.py
import paho.mqtt.client as mqtt
import struct

# MQTT Configuration
BROKER = "127.0.0.1"
PORT = 1883
TOPIC = "factory/sensor/speed"

# Scaling factor (from datasheet)
SCALING_FACTOR = 0.1

# Callback when a message is received
def on_message(client, userdata, message):
    # Decode payload (assuming it's hex)
    raw_data = message.payload.decode("utf-8")
    print(f"Raw data received: {raw_data}")
    
    # Extract speed (first 4 bytes)
    speed_raw = int(raw_data[0:8], 16)  # Convert hex to int
    speed = speed_raw * SCALING_FACTOR  # Apply scaling
    
    # Extract device status (next byte)
    device_status = int(raw_data[8:10], 16)
    
    # Extract OUT1 and OUT2 statuses (last byte)
    digital_status = int(raw_data[10:12], 16)
    out1_status = bool(digital_status & 0x01)
    out2_status = bool((digital_status >> 1) & 0x01)
    
    # Print processed data
    print(f"Speed: {speed} Hz, Device Status: {device_status}, OUT1: {out1_status}, OUT2: {out2_status}")

# Setup MQTT client
client = mqtt.Client()
client.on_message = on_message
client.connect(BROKER, PORT)
client.subscribe(TOPIC)

print("Listening for MQTT messages...")
client.loop_forever()
